### PR TITLE
speedy: short circuit precondition checking

### PR DESCRIPTION
**Pull request by robin-da**
Instead of evaluating all preconditions of templates and interfaces we
evaluate preconditions one by one and throw an exception as soon as we
find a failing one.

This implemented via a lazy SBCheckPrecond builtin and a foldleft
expression over the list of preconditions.

There would be another way of doing this, by implementing something
similar like the foldl continuation just for the precondition
checking, but I'm not quite sure whether this is superior to this
solution.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
@cocreature opening this for some questions. I started with this solution and then continued to transform it, such that the precondition checking uses its own continuation in the speedy machine. I got stuck trying to pass the preconditions without evaluating them first. Now I'm asking myself, whether a specialized continuation for preconditions is actually much better than this solution via foldl. What's your opinion?

**Review by cocreature — COMMENTED** [review]
Looks pretty sensible to me but I’d like @sofiafaro-da and @remyhaemmerle-da to take a look as well.

**Comment by cocreature** [comment]
Note that this is also only the speedy part. We also need to make sure we preserve the order from the Daml source code to speedy and test for that.

**Comment by robin-da** [comment]
I'm aware of the order issue, I'll address this in the next PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -722,17 +722,15 @@ private[lf] final class Compiler(
...
-        s.SEValue.EmptyList
-      )).to(ImmArray)
-    val preconds = s.SEApp(s.SEBuiltin(SBConsMany(precondsArray.length - 1)), precondsArray.toList)
-    // We check precondition in a separated builtin to prevent
-    // further evaluation of agreement, signatories, observers and key
-    // in case of failed precondition.
-    let(env2, SBCheckPrecond(tmplId)(env2.toSEVar(tmplArgPos), preconds)) { (_, env) =>
+      (Iterator(translateExp(env2, tmpl.precond)) ++ implementsPrecondsIterator).to(ImmArray)
+
+    val preconds = precondsArray.foldLeft[s.SExpr](s.SEValue(SUnit))((acc, precond) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:729]: (nit picking) 
```suggestion
    val preconds = precondsArray.foldLeft[s.SExpr](s.SEValue.SUnit)((acc, precond) =>
```
avoids an allocation.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:729]: I am fine with the fold. 
Just wondering out loud if using nested `let` is not better than nested `application`. 
@nickchapman-da ?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -879,25 +885,23 @@ private[lf] object SBuiltin {
...
-  final case class SBCheckPrecond(templateId: TypeConName) extends SBuiltinPure(2) {
+  final case class SBCheckPrecond(templateId: TypeConName) extends SBuiltinPure(3) {
     override private[speedy] def executePure(args: util.ArrayList[SValue]): SUnit.type = {
-      if (
-        !(getSList(args, 1).iterator.forall(_ match {
-          case SBool(b) => b
-          case otherwise => crash(s"type mismatch SBCheckPrecond: expected SBool, got $otherwise")
-        }))
-      )
+      val _ = getSUnit(args, 1)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:894]: (another nit picking) 
```suggestion
      discard(getSUnit(args, 1))
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice. Thanks.

**Review by cocreature — APPROVED** [review]
Nice, can we add a test for this? Probably the evaluation order trace-based tests are a good place.

**Comment by robin-da** [comment]
:+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (robin-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: yep, fixed it.

**Review by sofiafaro-da — APPROVED** [review]
Nice, thank you!

**Pull request merged**